prow/bugzilla/client.go

```diff
@@ -75,25 +78,77 @@ type Client interface {
 	RemovePullRequestAsExternalBug(id int, org, repo string, num int) (bool, error)
 	// GetAllClones returns all the clones of the bug including itself
 	// Differs from GetClones as GetClones only gets the child clones which are one level lower
-	GetAllClones(bug *Bug) ([]*Bug, error)
+	// GetAllClones will also fetch bugDetails and the root bug details in addition to clones
+	// since we want to parallelize all the HTTP calls
+	GetAllClones(bugID int) (clones []*Bug, bug *Bug, root *Bug, err error)
 	// GetRootForClone returns the original bug.
 	GetRootForClone(bug *Bug) (*Bug, error)
+	// GetFromCache will fetch the clones from the clones cache
+	GetFromCache(key int) ([]int, bool)
+	// SetInCache will push the cloneIDs associated with the bug to the cache
+	SetInCache(key int, value []int)
 }
 
 func NewClient(getAPIKey func() []byte, endpoint string) Client {
 	return &client{
-		logger:    logrus.WithField("client", "bugzilla"),
-		client:    &http.Client{},
-		endpoint:  endpoint,
-		getAPIKey: getAPIKey,
+		logger:      logrus.WithField("client", "bugzilla"),
+		client:      &http.Client{},
+		endpoint:    endpoint,
+		getAPIKey:   getAPIKey,
+		clonesCache: clonesCacheWithMutex{cache: map[int][]int{}},
 	}
 }
 
+type clonesCacheWithMutex struct {
+	cache map[int][]int
+	lock  sync.Mutex
+}
+
+func (cc *clonesCacheWithMutex) Get(key int) ([]int, bool) {
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+	clones, ok := cc.cache[key]
+	return clones, ok
+}
+
+func (cc *clonesCacheWithMutex) Set(key int, value []int) {
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+	cc.cache[key] = value
+}
+
+// NewBugDetailsCache is a constructor for bugDetailsCache
+func NewBugDetailsCache() *BugDetailsCache {
+	return &BugDetailsCache{cache: map[int]*Bug{}}
+}
+
+// BugDetailsCache holds the already retrieved bug details
+type BugDetailsCache struct {
+	cache map[int]*Bug
+	lock  sync.Mutex
+}
+
+// Get retrieves bug details from the cache and is thread safe
+func (bd *BugDetailsCache) Get(key int) (*Bug, bool) {
+	bd.lock.Lock()
+	defer bd.lock.Unlock()
+	bug, ok := bd.cache[key]
+	return bug, ok
+}
+
+// Set stores the bug details in the cache and is thread safe
+func (bd *BugDetailsCache) Set(key int, value *Bug) {
+	bd.lock.Lock()
+	defer bd.lock.Unlock()
+	bd.cache[key] = value
+}
+
 type client struct {
-	logger    *logrus.Entry
-	client    *http.Client
-	endpoint  string
-	getAPIKey func() []byte
+	logger      *logrus.Entry
+	client      *http.Client
+	endpoint    string
+	getAPIKey   func() []byte
+	clonesCache clonesCacheWithMutex
```

Comment:
Wait, is this currentl used?

Reply:
Yes, this is the cache used by the Get/Set function you had commented on earlier.